src/hotspot/share/gc/serial/serialStringDedup.hpp

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2021, Alibaba Group Holding Limited. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#ifndef SHARE_GC_SERIAL_STRINGDEDUP_HPP
+#define SHARE_GC_SERIAL_STRINGDEDUP_HPP
+
+#include "gc/shared/stringdedup/stringDedup.hpp"
+#include "memory/allStatic.hpp"
+#include "oops/oopsHierarchy.hpp"
+
+class SerialStringDedup : AllStatic {
+public:
+
+  // Candidate selection policy for full GC, returning true if the given
+  // String is a candidate for string deduplication.
+  // precondition: StringDedup::is_enabled()
+  // precondition: java_string is a Java String
+  static bool is_candidate_from_mark(oop java_string);
+
+  // Candidate selection policy for young during evacuation.
+  // If to is young then age should be the new (survivor's) age.
+  // if to is old then age should be the age of the copied from object.
+  static bool is_candidate_from_evacuation(oop java_string,
+                                           bool obj_is_tenured) {
+    return StringDedup::is_enabled() &&
+           java_lang_String::is_instance_inlined(java_string) &&
```

Comment:
Missing `#include` for this function.  A bunch of the pre-submit builds on github are failing because of this.  To call this function will require including javaClasses.inline.hpp, and ".inline.hpp" files cannot be included by an ordinary ".hpp".  Dealing with this will require adding "gc/serial/stringDedup.inline.hpp", moving this definition there, &etc.